### ISSUE TITLE
Explaining how to handle focus after clipboard actions

### ISSUE THREAD

**Issue by ZoeBijl** [w3c/html#166]
It is important to inform developers on how to manage focus with the use of the clipboard events. For example, where should focus move after a cut or paste command?

Propose to add text to [focus section in HTML 5.1](https://w3c.github.io/html/editing.html#focus) that explains how to handle focus in these situations.

[Original issue on the Clipboard API](https://github.com/w3c/clipboard-apis/issues/26).

**Comment by ZoeBijl** [comment]
I have a first (plain text) draft for this, where should I post that?

**Comment by chaals** [comment]
How does it compare to what browsers do? If you've got the pieces in place and its interoperable, you can make it into a pull request. If you're just thinking out loud, you probably want to post something here and refer to the Clipboard API or vice versa.

**Comment by chaals** [comment]
@MichielBijl can you propose a Pull Request in the next few weeks, to make the milestone?

**Comment by LJWatson** [comment]
@MichielBijl any progress on this?

**Comment by ZoeBijl** [comment]
Will post update today. Reading through section “5.4 Focus”.

Initial bit of research on where cursors are placed after a cut or paste action:

# Observations
Some notes on observations from various platforms.

## macOS 10.12
1. Select text with the keyboard (shift+arrows)
2. Cut text (cmd+x)
3. Cursor is placed in place of text

1. Select text with the mouse (double click on word)
2. Cut text (cmd+x)
3. If word/text is in the middle of a sentence surrounding spaces will collapse.
4. Cursor is placed in place of text

1. Paste object (cmd+v)
2. Cursor is placed at the end (right) of pasted object
	- In RTL languages it’s placed on the left
	- In TTB languages it’s placed at the bottom
3. If the word was selected with the mouse a space will be added if the paste cursor is placed against other text.

## Windows 10
1. Select text with the keyboard
2. Cut text (ctrl+x)
3. Cursor is placed in place of text

1. Select text with the mouse (double click on word)
2. Cut text (ctrl+x)
3. If word/text is in the middle of a sentence the space after the selected text will be included in the selection.
4. Cursor is placed in place of text

1. Paste object (ctrl+v)
2. Cursor is placed at the end of pasted object

## iOS 10 / Android 7
1. Select text (double tap)
2. Cut text (select cut from pop-up menu)
3. Cursor is placed in place of text

1. Paste object (tap anywhere in text → tap paste from pop-up menu)
2. Cursor is placed at end (right) of text

***

## Conclusion
The cursor is always placed towards the end of the sentence. For example:

-  When a user writes in a right-to-left language, the cursor will be placed to the left of the pasted object.
- In left-to-right languages the cursor is placed to the right of the pasted object.
- In top-to-bottom languages the cursor is placed at the bottom of the pasted object.

**Comment by ZoeBijl** [comment]
A draft for this section:

# Focus handling for clipboard actions

There are three clipboard actions that can affect focus: cut, copy, and paste. Focus for these actions is handled by the user agent by default. When custom clipboard events are implemented these default behaviours should be implemented by the author.

- Cut action
	- When a cut action is performed the selected object should be removed and focus is set in place of the cut object.
- Copy action
	- A copy action does not affect focus or selection. The selected object stay focused and selected.
- Paste action
	- When a paste action is performed focus is placed directly after the pasted object and towards the end of the writing direction.
	- For example in a right-to-left language the text cursor is placed to the left of the pasted object.

**Comment by chaals** [comment]
how about changing "towards the end of the writing direction" to "after the final character, according to the writing direction of that character"?

Otherwise, LGTM. Want to make a Pull Request?

**Comment by ZoeBijl** [comment]
@chaals how about “after the final character in the writing direction”?

If you can tell where you want this in the section then I can create a PR.

**Comment by chaals** [comment]
Sounds good. I figure you should put it where it makes most sense to you, and the editors can shift it if they really want it elsewhere. I find that entire section lacking in organisation.

If it were me I would make a 5.4.7 for it. And then try to find some order in the chaos…

**Comment by rniwa** [comment]
"after the pasted object and towards the end of the writing direction" is a bit ambiguous here. You should probably mention that "the collapsed selection should be placed at the *end* of the pasted contents." The [CSS Writing Modes Level 3](https://drafts.csswg.org/css-writing-modes-3/#start) clearly defines the meaning of *end* even in vertical writing mode whereas the notion of *after* isn't well defined in any spec as far as I know.

Since it's possible for an application to use either the paragraph level direction (e.g. the entire paragraph is LTR or RTL) or the inline direction of the pasted content (though the pasted content could contain both LTR & RTL text and the overall direction might be ambiguous at least to the end user), I don't think we can necessarily recommend one or the other in terms of whether inline-direction or paragraph-direction is used to determine which side is the "end".

**Comment by ZoeBijl** [comment]
@rniwa @chaals, what should we do with that sentence, as that is the only thing holding this back and it would be nice to have this in the update for the 20th.

My use of <q>towards the end of the writing direction</q> was based on CSS language. That said, I like rniwa’s suggestion: <q>the collapsed selection should be placed at the end of the pasted contents.</q>

So, maybe something like:

> When a paste action is performed focus should be placed at the end of the pasted contents.

**Comment by rniwa** [comment]
It's important to realize that the directionality wouldn't matter in terms of DOM offsets. As long as you place selection's start & end container & offset at the end of the pasted content regardless of whether the last few characters were RTL or LTR, the browsers would display the caret at the "right" location in accordance with their mapping between DOM offsets and visual positions (although Gecko can display the caret at two different locations depending on from which direction the caret has moved, and I don't think you can control this in JS).

Note that what it means to place the caret at the end of a text run is ambiguous, and the "correct" caret placement depends and differs on each platform. e.g. Windows uses keyboard's direction to determine whether it shows a caret using LTR convention or RTL convention, and Mac would, under the right user preferences, display a split caret at both locations although WebKit doesn't currently support that. For these reasons, I don't think we can mandate anything about where caret should be placed visually.

**Comment by LJWatson** [comment]
Thanks @rniwa. Could you propose an amendment to @michielbijl's text?

**Comment by rniwa** [comment]
I've done that in https://github.com/w3c/html/issues/166#issuecomment-279503130: "the collapsed selection should be placed at the end of the pasted contents."

**Comment by LJWatson** [comment]
Thanks @rniwa. @MichielBijl can you update the text and get a PR in for the editors to review today?

**Comment by ZoeBijl** [comment]
I had some difficulty with understanding “collapsed selection”, but after some reading up I figured it out, it simply means the caret.

Perhaps it would be beneficial to add this to the glossary for other people that aren’t familiar with the term.

Another question I’d like to ask, I used focus in all my wording because we could also be dealing with files in a desktop like or file management environment; if you paste a file to say Dropbox, focus should be set on the pasted file. Maybe it’s better to make that a separate bullet?

@rniwa said:

> For these reasons, I don't think we can mandate anything about where caret should be placed visually.

Does this mean you suggest we remove this line: “For example in a right-to-left language the text cursor is placed to the left of the pasted object.”?

***

I have this all in my fork, so can create a PR once this issues are resolved.

**Comment by rniwa** [comment]
> Does this mean you suggest we remove this line: “For example in a right-to-left language the text cursor is placed to the left of the pasted object.”?

Yes. For starters, that sentence is misnomer because RTL languages such as Hebrew and Arabic use both LTR and RTL characters; e.g. numbers. So it's not obvious where the caret should be placed visually relative to the pasted content. In some cases, it may need to appear within the paste content.

For example, if you had three RTL letters, A, B, C, and then a number 1, 2, 3 in the document/logical order and the content was pasted in a RTL block, it would visually look like 123ABC.  At this point, you could argue that the caret should be placed between 3 and A because that's the place a new LTR number 4 should appear if typed, or on the left of 1 because that's where another RTL letter such as D would appear if typed.

There is no agreed upon convention across different platforms and text processing systems.

**Comment by rniwa** [comment]
> Another question I’d like to ask, I used focus in all my wording because we could also be dealing with files in a desktop like or file management environment; if you paste a file to say Dropbox, focus should be set on the pasted file. Maybe it’s better to make that a separate bullet?

Right, we need to disambiguate the two. Focus is a completely different concept from selection. In particular, in WebKit or Mac in general, moving selection does not automatically move focus per platform convention.

**Comment by ZoeBijl** [comment]
I clearly need to work on this more, although I don’t see that happening before the 24th, I’m sorry.

**Comment by arronei** [comment]
@MichielBijl I assigned this to you to finish up the PR. If you could get that ready soon I will take a look at it and hopefully get this item done.
